File: Gombang/[LeetCode75] - [Priority_Queue]/2336_Smallest_Number_in_Infinite_Set.cpp
// 모범 풀이법.
// currNum이라는 값을 통해 양의 정수를 직접 표현.
// 중간중간에 currNum보다 작은 값이 들어오면 set에 데이터 저장.
class SmallestInfiniteSet {
public:

    SmallestInfiniteSet() {
        currNum = 1;
    }

    int popSmallest() {

        if (infiniteSet.size() > 0)
        {
            int smallestNumber = *infiniteSet.begin();
            infiniteSet.erase(infiniteSet.begin());
            return smallestNumber;
        }
        else
        {
            return currNum++;
        }
    }

    void addBack(int num) {

        if (num >= currNum)
            return;

        infiniteSet.insert(num);

    }

private:
    // 오름차순 정렬을 자동으로 지원해주기 때문에, 가장 앞에 있는 값이 최소값을 보장.
    set<int> infiniteSet;

    // 양의 정수를 담당하는 변수.
    int currNum;
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */

// 첫 번째 풀이.
// 1부터 1000이라는 값을 그대로 넣어줘서 진행.
// 불필요한 코드들이 많았음.
class SmallestInfiniteSet {
public:
    SmallestInfiniteSet() {
        // At most 1000 calls 라는 것이 있어서 1000개의 데이터를 넣어줌.
        for (int i = 1; i <= 1000; i++)
        {
            infiniteSet.insert(i);
            minHeap.push(i);
        }
    }

    int popSmallest() {
        int minValue = minHeap.top();
        minHeap.pop();

        infiniteSet.erase(minValue);
        return minValue;
    }

    void addBack(int num) {
        if (infiniteSet.find(num) != infiniteSet.end())
            return;

        infiniteSet.insert(num);
        minHeap.push(num);
    }

private:
    priority_queue<int, vector<int>, greater<int>> minHeap;
    unordered_set<int> infiniteSet;
};

File: Lucy/2025/4-W04/2336_Smallest_Number_in_Infinite_Set.ts
import { MinPriorityQueue } from "@datastructures-js/priority-queue";

// 31 ms
class SmallestInfiniteSet {
  private minHeap: MinPriorityQueue<number>;
  private nextSmallest: number;

  private seen: Set<number>;

  constructor() {
    this.minHeap = new MinPriorityQueue<number>();
    this.nextSmallest = 1;
    this.seen = new Set<number>();
  }

  popSmallest(): number {
    if (!this.minHeap.isEmpty()) {
      const popped = this.minHeap.dequeue();
      if (popped) {
        this.seen.delete(popped);

        return popped;
      }
    }

    return this.nextSmallest++;
  }

  addBack(num: number): void {
    if (this.nextSmallest > num && !this.seen.has(num)) {
      this.minHeap.enqueue(num);
      this.seen.add(num);
    }
  }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * var obj = new SmallestInfiniteSet()
 * var param_1 = obj.popSmallest()
 * obj.addBack(num)
 */

// 모범 답안 25ms
class SmallestInfiniteSet2 {
  private heap: number[];
  private seen: Set<number>;
  private current: number;

  constructor() {
    this.heap = [];
    this.seen = new Set();
    this.current = 1;
  }

  heapifyUp(): void {
    let i = this.heap.length - 1;

    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);

      if (this.heap[parent] > this.heap[i]) {
        [this.heap[i], this.heap[parent]] = [this.heap[parent], this.heap[i]];
        i = parent;
      } else {
        break;
      }
    }
  }

  heapifyDown(): void {
    const last = this.heap.pop();

    if (this.heap.length > 0 && last) {
      this.heap[0] = last;

      let i = 0;
      while (true) {
        const left = i * 2 + 1;
        const right = i * 2 + 2;

        let s = i;
        if (left < this.heap.length && this.heap[left] < this.heap[s]) s = left;
        if (right < this.heap.length && this.heap[right] < this.heap[s]) s = right;
        if (s === i) break;

        [this.heap[s], this.heap[i]] = [this.heap[i], this.heap[s]];

        i = s;
      }
    }
  }

  popSmallest(): number {
    if (this.heap.length > 0) {
      const smallest = this.heap[0];
      this.heapifyDown();

      this.seen.delete(smallest);
      return smallest;
    }
    return this.current++;
  }

  addBack(num: number): void {
    if (num < this.current && !this.seen.has(num)) {
      this.seen.add(num);

      this.heap.push(num);

      this.heapifyUp();
    }
  }
}
